Redo jit/type and utils/functional to ATen/core

bwasti opened the pull request:
Summary: This is a redo of the previous move which broke OS X and Windows tests -- RTTI seemed to be broken

Differential Revision: D12883775

ezyang commented:
cc @zdevito

Yangqing reviewed (approved):
(stamping per request)

bwasti committed:
Redo jit/type and utils/functional to ATen/core (#13455)

Summary:
This is a redo of the previous move which broke OS X and Windows tests -- RTTI seemed to be broken
Pull Request resolved: https://github.com/pytorch/pytorch/pull/13455

Differential Revision: D12883775

fbshipit-source-id: d2f0a86a1358bce98a1c74c5230ff1dcb2a22b46
